subprojects/wrapper/src/integTest/groovy/org/gradle/integtests/WrapperLoggingIntegrationTest.groovy

```diff
@@ -23,27 +23,59 @@ import org.gradle.util.TestPrecondition
 class WrapperLoggingIntegrationTest extends AbstractWrapperIntegrationSpec {
 
     def setup() {
-        executer.withWelcomeMessageEnabled()
         file("build.gradle") << "task emptyTask"
     }
 
-    def "wrapper only renders welcome message when executed in quiet mode"() {
+    def "wrapper does not render welcome message when executed in quiet mode"() {
         given:
         prepareWrapper()
+        executer.withWelcomeMessageEnabled()
 
         when:
         args '-q'
         result = wrapperExecuter.withTasks("emptyTask").run()
 
+        then:
+        result.output.empty
+    }
+
+    def "wrapper renders welcome message when executed the first time"() {
+        given:
+        prepareWrapper()
+
+        when:
+        executer.withWelcomeMessageEnabled()
+        result = wrapperExecuter.withTasks("emptyTask").run()
+
         then:
         outputContains("Welcome to Gradle $wrapperExecuter.distribution.version.version!")
 
         when:
+        executer.withWelcomeMessageEnabled()
+        result = wrapperExecuter.withTasks("emptyTask").run()
+
+        then:
+        outputDoesNotContain("Welcome to Gradle $wrapperExecuter.distribution.version.version!")
+    }
+
+    def "wrapper renders welcome message when executed the first time after being executed in quiet mode"() {
+        given:
+        prepareWrapper()
+
+        when:
+        executer.withWelcomeMessageEnabled()
         args '-q'
         result = wrapperExecuter.withTasks("emptyTask").run()
```

Comment:
I think this test never really worked, because `run()` calls reset on the gradle executer which sets the `welcomeMessageEnabled` flag to false. So it did not test whether the marker file handling works because the welcome message wouldn't be rendered anyway. For this reason I explicitly call `withWelcomeMessageEnabled()` everywhere.

Reply:
:x: Good catch! You can use the following in `setup()` instead of repeating the method call:

```groovy
executer.beforeExecute {
    withWelcomeMessageEnabled()
}
```

Reply:
Done